runner/watcher.go

```diff
@@ -34,25 +34,38 @@ func (lw *logWatcher) findStreams() ([]*string, error) {
 	return streams, err
 }
 
-func (lw *logWatcher) printEventsAfter(ts int64, c context.CancelFunc) (int64, error) {
-	var streams []*string
-	var err error
-
-	log.Printf("Printing events after %d", ts)
+func (lw *logWatcher) waitForSteams(ctx context.Context) ([]*string, error) {
+	for {
+		select {
+		case <-time.After(2 * time.Second):
+			streams, err := lw.findStreams()
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == "Throttling" {
+					time.Sleep(5 * time.Second)
+					continue
+				}
+			}
+			if err != nil {
+				return nil, err
+			}
+			if len(streams) > 0 {
+				return streams, nil
+			}
 
-	// sometimes the stream takes a while to appear
-	log.Printf("Waiting for log stream to start...")
-	for i := 0; i < 300; i++ {
-		streams, err = lw.findStreams()
-		if err != nil || len(streams) == 0 {
-			time.Sleep(time.Second * 2)
+		case <-ctx.Done():
+			return nil, ctx.Err()
 		}
 	}
+}
 
+func (lw *logWatcher) printEventsAfter(ctx context.Context, ts int64, c context.CancelFunc) (int64, error) {
+	log.Printf("Printing events after %d", ts)
+
+	// sometimes the stream takes a while to appear
+	log.Print("Waiting for log stream to start...")
+	streams, err := lw.waitForSteams(ctx)
```

Comment:
Don't forget that you will want a timeout for this context too! Otherwise it will wait forever for streams to appear.

Reply:
Alternately, a `case <-time.After(1 * time.Minute):` would do the trick for giving up.

Reply:
The context being used here has already been given a timeout in [the calling Watch function](https://github.com/buildkite/ecs-run-task/pull/10/files#diff-047c56836677c48bae29a5bf4373e07eR95) - is that enough? The context [is being checked in waitForStreams](https://github.com/buildkite/ecs-run-task/pull/10/files?diff=split#diff-047c56836677c48bae29a5bf4373e07eR55)

Reply:
You are only creating a timeout context AFTER `waitForStreams`. At the minute there is no mechanism for timing out `waitForStreams`. Is that your intention?

Reply:
I don't see it that way - `Watch` [creates a child context with a 5 minute timeout](https://github.com/buildkite/ecs-run-task/pull/10/files?diff=unified#diff-047c56836677c48bae29a5bf4373e07eR95). `Watch` [calls in to `printEventsAfter` with that context that has a timeout](https://github.com/buildkite/ecs-run-task/pull/10/files?diff=unified#diff-047c56836677c48bae29a5bf4373e07eR101). `printEventsAfter` calls `waitForStreams` on this line with that same context that has the timeout. This would mean the timeout context is being created before `waitForStreams`.

Reply:
Yup, I see what you are saying. The issue with that is it's really hard to figure that out. If you create a child context specifically within `waitForSteams`, it's immediately obvious what it's for. If the parent context cancels, the child context will too. That's generally how I approach contexts, otherwise it's breaking method encapsulation IMO.

Reply:
That said, I don't practice what I preach in parfait! 🤷🏼‍♂️